hbase-client/src/main/java/org/apache/hadoop/hbase/filter/PrefixFilter.java

```diff
@@ -43,18 +43,25 @@ public class PrefixFilter extends FilterBase implements HintingFilter {
   protected boolean filterRow = true;
   protected boolean provideHint = false;
   protected Cell reversedNextCellHint;
+  protected Cell forwardNextCellHint;
 
   public PrefixFilter(final byte[] prefix) {
     this.prefix = prefix;
-    // Pre-compute reverse hint at creation to avoid re-computing it several times in the corner
+    // Pre-compute hints at creation to avoid re-computing them several times in the corner
     // case where there are a lot of cells between the hint and the first real match.
-    this.reversedNextCellHint = createReversedNextCellHint();
+    createCellHints();
   }
 
-  private Cell createReversedNextCellHint() {
+  private void createCellHints() {
+    if (prefix == null) {
+      return;
+    }
     // On reversed scan hint should be the prefix with last byte incremented
     byte[] reversedHintBytes = increaseLastNonMaxByte(this.prefix);
-    return PrivateCellUtil.createFirstOnRow(reversedHintBytes, 0, (short) reversedHintBytes.length);
+    this.reversedNextCellHint =
```

Comment:
I was about to say that we could choose the director here, but interestingly the direction is not available here yet.

Reply:
Practically, I'm sure it's not changed after any of the Cell processing methods have been called.
It is called once when the Scan is set up. (perhaps also after reset() is called)